tracing-flame/src/lib.rs

```diff
@@ -0,0 +1,397 @@
+//! ```
+//!
+//! As an alternative, you can provide _any_ type that implements `std::io::Write` to
+//! `FlameLayer::new`.
+//!
+//! ## Generating The Image
+//!
+//! To convert the textual representation of a flamegraph to a visual one, first install `inferno`:
+//!
+//! ```console
+//! cargo install inferno
+//! ```
+//!
+//! Then, pass the file created by `FlameLayer` into `inferno-flamegraph`:
+//!
+//! ```console
+//! # flamegraph
+//! cat tracing.folded | inferno-flamegraph > tracing-flamegraph.svg
+//!
+//! #flamechart
+//! cat tracing.folded | inferno-flamegraph --flamechart > tracing-flamechart.svg
+//! ```
+//!
+//! ## Differences between `flamegraph`s and `flamechart`s
+//!
+//! By default, `inferno-flamegraph` creates flamegraphs. Flamegraphs operate by
+//! that collapsing identical stack frames and sorting them on the frame's names.
+//!
+//! This behavior is great for multithreaded programs and long-running programs
+//! where the same frames occur _many_ times, for short durations, because it reduces
+//! noise in the graph and gives the reader a better idea of the
+//! overall time spent in each part of the application.
+//!
+//! However, it is sometimes desirable to preserve the _exact_ ordering of events
+//! as they were emitted by `tracing-flame`, so that it is clear when each
+//! span is entered relative to others and get an accurate visual trace of
+//! the execution of your program. This representation is best created with a
+//! `flamechart`, which _does not_ sort or collapse identical stack frames.
+//!
+//! [`tracing`]: https://docs.rs/tracing
+//! [`inferno`]: https://docs.rs/inferno
+//! [`FlameLayer`]: struct.FlameLayer.html
+//! [`FlushGuard`]: struct.FlushGuard.html
+#![warn(
+    missing_debug_implementations,
+    missing_docs,
+    rust_2018_idioms,
+    unreachable_pub,
+    bad_style,
+    const_err,
+    dead_code,
+    improper_ctypes,
+    non_shorthand_field_patterns,
+    no_mangle_generic_items,
+    overflowing_literals,
+    path_statements,
+    patterns_in_fns_without_body,
+    private_in_public,
+    unconditional_recursion,
+    unused,
+    unused_allocation,
+    unused_comparisons,
+    unused_parens,
+    while_true
+)]
+
+pub use error::Error;
+
+use error::Kind;
+use std::cell::Cell;
+use std::fmt;
+use std::fmt::Write as _;
+use std::fs::File;
+use std::io::BufWriter;
+use std::io::Write;
+use std::marker::PhantomData;
+use std::path::Path;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::time::{Duration, Instant};
+use tracing::span::Attributes;
+use tracing::Id;
+use tracing::Subscriber;
+use tracing_subscriber::layer::Context;
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::registry::SpanRef;
+use tracing_subscriber::Layer;
+
+mod error;
+
+thread_local! {
+    static START: Instant = Instant::now();
+    static LAST_EVENT: Cell<Instant> =  Cell::new(START.with(|f| *f));
+}
+
+/// A Layer that logs span open/close events as folded flame graph stack
```

Comment:
nit:
```suggestion
/// A `Layer` that records span open/close events as folded flamegraph stack
```
and maybe also link to the `Layer` trait?